### ISSUE TITLE
libuv: Connection / disconnection mismatch

### ISSUE THREAD

**Issue by alexhultman** [warmcat/libwebsockets#448]
Hi.

I've found an example that reproduces the problems with the connection / disconnection unbalance. So if you send an upgrade request and then terminate the sender before the connection has been established in libwebsockets you will not get a close event triggered. The connection will be "stuck" forever in libwebsockets. This applies to 1.7 stable with the uv server (maybe others too).

```
#include <unistd.h>
#include <stdio.h>
#include <arpa/inet.h>
#include <sys/socket.h>

int main()
{
    int socketfd = socket(AF_INET, SOCK_STREAM, IPPROTO_TCP);
    struct sockaddr_in addr = {};
    addr.sin_family = AF_INET;
    addr.sin_addr.s_addr = inet_addr("127.0.0.1");
    addr.sin_port = htons(3000);

    char buf[] = "GET /default HTTP/1.1\r\n"
                "Host: server.example.com\r\n"
                "Upgrade: websocket\r\n"
                "Connection: Upgrade\r\n"
                "Sec-WebSocket-Key: x3JJHMbDL1EzLkh9GBhXDw==\r\n"
                "Sec-WebSocket-Protocol: default\r\n"
                "Sec-WebSocket-Version: 13\r\n"
                "Origin: http://example.com\r\n\r\n";

    connect(socketfd, (struct sockaddr *) &addr, sizeof(addr));
    send(socketfd, buf, sizeof(buf), 0);
}
```

This is on localhost so the server should get a notice about the closed socket. If you run this code it will tigger one connection but no disconnection.

**Comment by lws-team** [comment]
"The problems".... I didn't notice you mention a problem like that until now.

Checking your app (with some return checking and final return 0;) with the lws test apps on master -->

It doesn't make trouble on normal poll()-based event loop

```
lwsts[19119]: Accepted 0x15bfeb0 to tsi 0
lwsts[19119]: lws_header_table_attach: wsi 0x15bfeb0: ah (nil) (tsi 0)
lwsts[19119]: lws_header_table_attach: wsi 0x15bfeb0: ah 0x1594150: count 1 (on exit)
lwsts[19119]: Upgrade to ws
lwsts[19119]: checking default
lwsts[19119]: No protocol from "default" supported
lwsts[19119]: lws_header_table_detach: wsi 0x15bfeb0: ah 0x1594150 (tsi=0, count = 1)
lwsts[19119]: lws_close_free_wsi: shutting down connection: 0x15bfeb0
lwsts[19119]: wsi 0x15bfeb0: TIMEDOUT WAITING on 13 (did hdr 1, ah (nil), wl 0, pfd events 0)
lwsts[19119]: lws_close_free_wsi: real just_kill_connection: 0x15bfeb0
lwsts[19119]: remove_wsi_socket_from_fds: wsi=0x15bfeb0, sock=8, fds pos=2, end guy pos=3, endfd=0
lwsts[19119]: lws_header_table_detach: wsi 0x15bfeb0: ah (nil) (tsi=0, count = 0)
```

On libev test app, the close timeout is deferred until something else happens in the libev event loop, like another connection.  I think it's because I didn't add a native timeout "pulse" like I did for libuv, but it doesn't sound so terrible.

On libuv, it doesn't seem to make trouble either.  The below all happens at once.

```
lwsts[21398]: Accepted 0xcd07b0 to tsi 0
lwsts[21398]: lws_header_table_attach: wsi 0xcd07b0: ah (nil) (tsi 0)
lwsts[21398]: lws_header_table_attach: wsi 0xcd07b0: ah 0xc9f450: count 1 (on exit)
lwsts[21398]: Upgrade to ws
lwsts[21398]: checking default
lwsts[21398]: No protocol from "default" supported
lwsts[21398]: lws_header_table_detach: wsi 0xcd07b0: ah 0xc9f450 (tsi=0, count = 1)
lwsts[21398]: lws_close_free_wsi: shutting down connection: 0xcd07b0
lwsts[21398]: lws_close_free_wsi: real just_kill_connection: 0xcd07b0
lwsts[21398]: remove_wsi_socket_from_fds: wsi=0xcd07b0, sock=12, fds pos=1, end guy pos=2, endfd=0
lwsts[21398]: lws_header_table_detach: wsi 0xcd07b0: ah (nil) (tsi=0, count = 0)
```

...am I missing your point?

**Comment by alexhultman** [comment]
```lwsts[21398]: No protocol from "default" supported``` <- are you sure youre getting the connection event? Because when I run this I get a connection event (I have a "default" protocol) but I get no close event:

LWS_CALLBACK_ESTABLISHED but no LWS_CALLBACK_CLOSED

**Comment by alexhultman** [comment]
If you change the protocol in the header to "lws-mirror-protocol" or anything supported and add printf's to LWS_CALLBACK_ESTABLISHED and LWS_CALLBACK_CLOSED (or breakpoints or whatever) you'll get a connection but no disconnection.

So when I test this against the wrapper I can get it to just keep on gaining connections even though they are all closed.

**Comment by lws-team** [comment]
I tested it with your test app on the official test apps... that's the only sane way I think.

I have to do other things but can you try reproduce this on the test apps?  It sounds like just set the protocol name to one supported by the test apps is enough.

**Comment by alexhultman** [comment]
I have reproduced it in the test apps. They give me the connection but then they also seem to hang or something, I cannot get them to accept more connections.

**Comment by alexhultman** [comment]
This is what I get (with added printf's):
```
[alexhultman@localhost libwebsockets]$ bin/libwebsockets-test-server-libuv 
lwsts[17315]: libwebsockets test server libuv - license LGPL2.1+SLE
lwsts[17315]: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
lwsts[17315]: Initial logging level 7
lwsts[17315]: Libwebsockets version: 1.7.0 alexhultman@localhost.localdomain-v1.7.0-52-ge3d141d
lwsts[17315]: IPV6 not compiled in
lwsts[17315]: libev support not compiled in
lwsts[17315]:  Threads: 1 each 1500000 fds
lwsts[17315]:  mem: platform fd map: 12000000 bytes
lwsts[17315]:  mem: per-conn:          584 bytes + protocol rx buf
lwsts[17315]:  canonical_hostname = localhost.localdomain
lwsts[17315]:  Compiled with OpenSSL support
lwsts[17315]:  Using non-SSL mode
lwsts[17315]:  SSL ECDH curve 'prime256v1'
lwsts[17315]:  Listening on port 7681
    get  = /default
    host: = server.example.com
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://example.com
    sec-websocket-protocol: = dumb-increment-protocol
    http/1.1  = HTTP/1.1
    sec-websocket-key: = x3JJHMbDL1EzLkh9GBhXDw==
    sec-websocket-version: = 13
Connection
```

Then I cannot connect any more. When I terminate the server it prints this:

```
lwsts[17315]: Signal 2 caught, exiting...
Close
lwsts[17315]: lws_header_table_detach: wsi 0xd3a760: ah held 57s, ah.rxpos 242, ah.rxlen 242, mode/state 3 4,wsi->more_rx_waiting 1
lwsts[17315]: lws_context_destroy
lwsts[17315]: callback_lws_mirror: mirror protocol cleaning up
lwsts[17315]: libwebsockets-test-server exited cleanly
```

Where "Connection" and "Close" are my own added printf's.

**Comment by alexhultman** [comment]
It works with the non-uv server, there it closes directly and I can perform more connections:

```
[alexhultman@localhost libwebsockets]$ bin/libwebsockets-test-server
lwsts[17354]: libwebsockets test server - license LGPL2.1+SLE
lwsts[17354]: (C) Copyright 2010-2016 Andy Green <andy@warmcat.com>
Using resource path "/usr/share/libwebsockets-test-server"
lwsts[17354]: Initial logging level 7
lwsts[17354]: Libwebsockets version: 1.7.0 alexhultman@localhost.localdomain-v1.7.0-52-ge3d141d
lwsts[17354]: IPV6 not compiled in
lwsts[17354]: libev support not compiled in
lwsts[17354]:  Threads: 1 each 1500000 fds
lwsts[17354]:  mem: platform fd map: 12000000 bytes
lwsts[17354]:  mem: per-conn:          584 bytes + protocol rx buf
lwsts[17354]:  canonical_hostname = localhost.localdomain
lwsts[17354]:  Compiled with OpenSSL support
lwsts[17354]:  SSL ciphers: 'ECDHE-ECDSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-GCM-SHA384:DHE-RSA-AES256-GCM-SHA384:ECDHE-RSA-AES256-SHA384:HIGH:!aNULL:!eNULL:!EXPORT:!DES:!MD5:!PSK:!RC4:!HMAC_SHA1:!SHA1:!DHE-RSA-AES128-GCM-SHA256:!DHE-RSA-AES128-SHA256:!AES128-GCM-SHA256
lwsts[17354]:  Using non-SSL mode
lwsts[17354]:  SSL ECDH curve 'prime256v1'
lwsts[17354]:  Listening on port 7681
    get  = /default
    host: = server.example.com
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://example.com
    sec-websocket-protocol: = dumb-increment-protocol
    http/1.1  = HTTP/1.1
    sec-websocket-key: = x3JJHMbDL1EzLkh9GBhXDw==
    sec-websocket-version: = 13
Connection
lwsts[17354]: lws_service_fd_tsi: 0x23240b0: using inherited ah rx
lwsts[17354]: lws_service_fd_tsi: 0x23240b0: detaching inherited used ah
Close
    get  = /default
    host: = server.example.com
    connection: = Upgrade
    upgrade: = websocket
    origin: = http://example.com
    sec-websocket-protocol: = dumb-increment-protocol
    http/1.1  = HTTP/1.1
    sec-websocket-key: = x3JJHMbDL1EzLkh9GBhXDw==
    sec-websocket-version: = 13
Connection
lwsts[17354]: lws_service_fd_tsi: 0x2324c40: using inherited ah rx
lwsts[17354]: lws_service_fd_tsi: 0x2324c40: detaching inherited used ah
Close
```

**Comment by lws-team** [comment]
I am going to be working on something else all night.

libuv is pretty spartan with the event generation compared to anything else.  However if the underlying connection did actually close, check if we got the same situation we had in windows, POLLIN is enabled but no equivalent of POLLERR or POLLHUP or whatever libuv has.

**Comment by alexhultman** [comment]
I tried the libev server and that one works so it is only the uv server that doesn't work. So maybe I can solve this directly in libuv.c.

**Comment by alexhultman** [comment]
Yes you're right it is the errors that is not being handled correctly:
```
	if (status < 0)
		revents=UV_READABLE;//return;
```

That fixes it (but I should look into this some more to be sure about it).

**Comment by lws-team** [comment]
Thanks. Then I applied the PR, I assume we can close this.
